osu.Game/Overlays/Pause/PauseOverlay.cs

```diff
@@ -0,0 +1,214 @@
+﻿using System;
+using OpenTK;
+using OpenTK.Input;
+using OpenTK.Graphics;
+using osu.Game.Graphics;
+using osu.Framework.Input;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Transformations;
+using System.Threading.Tasks;
+
+namespace osu.Game.Overlays.Pause
+{
+    public class PauseOverlay : OverlayContainer
+    {
+        private const int transitionDuration = 200;
+        private const int buttonHeight = 70;
+        private const float backgroundAlpha = 0.75f;
+
+        public Action OnResume;
+        public Action OnRetry;
+        public Action OnQuit;
+
+        private FlowContainer retryCounterContainer;
+
+        public override bool Contains(Vector2 screenSpacePos) => true;
+        public override bool HandleInput => State == Visibility.Visible;
+
+        protected override void PopIn() => FadeIn(transitionDuration, EasingTypes.In);
+        protected override void PopOut() => FadeOut(transitionDuration, EasingTypes.In);
+
+        protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
+        {
+            switch (args.Key)
+            {
+                case Key.Escape:
+                    if (State == Visibility.Hidden) return false;
+                    Hide();
```

Comment:
make a method `resume() { Hide; OnResume?.Invoke }` as this is called in two different places.